transmission_interface/src/transmission_parser.cpp

```diff
@@ -12,68 +12,250 @@
 // limitations under the License.
 
-#include "transmission_interface/transmission_parser.hpp"
-#include "transmission_interface/transmission_info.hpp"
+#include <transmission_interface/transmission_parser.hpp>
 
-#include <tinyxml2.h>
-#include <stdexcept>
+// ROS
+#include <rclcpp/rclcpp.hpp>
+
+#include <sstream>
 #include <string>
 #include <vector>
 
-namespace
-{
-constexpr const auto kTransmissionTag = "transmission";
-constexpr const auto kJointTag = "joint";
-constexpr const auto kHardwareInterfaceTag = "hardwareInterface";
-}  // namespace
 
 namespace transmission_interface
 {
+static const rclcpp::Logger g_logger(rclcpp::get_logger("transmission_parser"));
 
 std::vector<TransmissionInfo> parse_transmissions_from_urdf(const std::string & urdf)
 {
-  if (urdf.empty()) {
-    throw std::runtime_error("empty URDF passed in to transmission parser");
+  TransmissionParser parser;
+  std::vector<TransmissionInfo> transmissions;
+  if (!parser.parse(urdf, transmissions)) {
+    throw std::runtime_error("failed to parse transmissions in URDF");
   }
+  return transmissions;
+}
+
 
-  tinyxml2::XMLDocument doc;
+bool TransmissionParser::parse(
+  const std::string & urdf,
+  std::vector<TransmissionInfo> & transmissions)
+{
+  // initialize TiXmlDocument doc with a string
+  TiXmlDocument doc;
   if (!doc.Parse(urdf.c_str()) && doc.Error()) {
-    throw std::runtime_error("invalid URDF passed in to transmission parser");
+    RCLCPP_ERROR(g_logger, "Can't parse transmissions. Invalid robot description.");
+    return false;
   }
 
-  std::vector<TransmissionInfo> transmissions;
   // Find joints in transmission tags
-  const tinyxml2::XMLElement * root_it = doc.RootElement();
-  const tinyxml2::XMLElement * trans_it = root_it->FirstChildElement(kTransmissionTag);
-  while (trans_it) {
+  TiXmlElement * root = doc.RootElement();
+
+  // Constructs the transmissions by parsing custom xml.
+  TiXmlElement * trans_it = nullptr;
+  for (trans_it = root->FirstChildElement("transmission"); trans_it;
+    trans_it = trans_it->NextSiblingElement("transmission"))
+  {
+    transmission_interface::TransmissionInfo transmission;
+
+    // Transmission name
+    if (trans_it->Attribute("name")) {
+      transmission.name = trans_it->Attribute("name");
+      if (transmission.name.empty()) {
+        RCLCPP_ERROR_STREAM(g_logger, "Empty name attribute specified for transmission.");
+        throw std::runtime_error("Empty name attribute specified for transmission");
+      }
+    } else {
+      RCLCPP_ERROR_STREAM(g_logger, "No name attribute specified for transmission.");
+      throw std::runtime_error("No name attribute specified for transmission");
+    }
+
+    // Transmission type
+    TiXmlElement * type_child = trans_it->FirstChildElement("type");
+    if (!type_child) {
+      RCLCPP_ERROR_STREAM(
+        g_logger, "No type element found in transmission '" <<
+          transmission.name << "'.");
```

Comment:
What do you think about:
```suggestion
      RCLCPP_ERROR_STREAM(
        g_logger,
        "No type element found in transmission '" << transmission.name << "'.");
```